File: src/localConsensusByN.py
from typing import List


def localConsensusByN(buffer: List[str], n: int) -> List[str]:
    # returnVals
    result: List[str] = []
    ############################ \returnVals

    
    
    # Check if buffer is not empty and contains lists
    if not buffer or not all(isinstance(lst, list) for lst in buffer):
        return result
    # Find the minimum length of the lists in buffer to avoid index errors
    # min_length = min(len(lst) for lst in buffer)
    
    max_length = max(len(lst) for lst in buffer)
    # Iterate over the range of the shortest list
    # for i in range(min_length):
    for i in range(max_length):
        # Create a dictionary to count occurrences of each token
        token_counts = {}
        # Iterate over each list in the buffer
        for inferenceTokens in buffer:
            # Standardize the token by lowercasing and stripping punctuation
            
            # if were off the end of this token buffer move on..
            if i >= len(inferenceTokens):
                continue
            
            token = inferenceTokens[i].lower().rstrip('.,')
            # Increment the token count
            token_counts[token] = token_counts.get(token, 0) + 1
        # Find tokens with a count of at least two and append to result
        for token, count in token_counts.items():
            if count >= n:
                result.append(token.rstrip('\'\"'))
                break  # Only one consensus token per index

    return result

File: src/cpp/localConsensusByN.cpp
#include "localConsensusByN.hpp"
#include <unordered_map>
#include <algorithm>

std::vector<std::string> localConsensusByN(const std::vector<std::vector<std::string>>& buffer, int n) {
    std::vector<std::string> result;
    if (buffer.empty()) {
        return result;
    }

    size_t max_length = 0;
    for (const auto& lst : buffer) {
        max_length = std::max(max_length, lst.size());
    }

    for (size_t i = 0; i < max_length; ++i) {
        std::unordered_map<std::string, int> token_counts;
        for (const auto& inferenceTokens : buffer) {
            if (i >= inferenceTokens.size()) {
                continue;
            }

            std::string token = inferenceTokens[i];
            std::transform(token.begin(), token.end(), token.begin(), ::tolower);
            token.erase(std::remove(token.end() - 2, token.end(), '.'), token.end());
            token.erase(std::remove(token.end() - 2, token.end(), ','), token.end());

            ++token_counts[token];
        }

        for (const auto& [token, count] : token_counts) {
            if (count >= n) {
                result.push_back(token.substr(0, token.size() - std::min(token.size(), static_cast<size_t>(2))));
                break;
            }
        }
    }

    return result;
}